File: checkpoints/checkpoint_20251006_140102_start_working_on_the_logbook_review_process_where_the_lgbooks_flows_back_and_forward_between_the_intern_and_supervisor_until_approved/code/backend/utils/duration_utils.py
"""
Utility functions for duration formatting and conversion
"""

def minutes_to_hours_minutes(minutes):
    """
    Convert minutes to hh:mm format
    Args:
        minutes: Duration in minutes (int or float)
    Returns:
        Formatted string like "3:45" or "0:30"
    """
    if not minutes or minutes < 0:
        return "0:00"
    
    # Round to avoid floating point precision issues
    total_minutes = round(minutes)
    hours = total_minutes // 60
    mins = total_minutes % 60
    
    return f"{hours}:{mins:02d}"

def minutes_to_display_format(minutes):
    """
    Convert minutes to display format (e.g., "3h 45m" or "30m")
    Args:
        minutes: Duration in minutes (int or float)
    Returns:
        Formatted string for display
    """
    if not minutes or minutes < 0:
        return "0m"
    
    total_minutes = int(minutes)
    hours = total_minutes // 60
    mins = total_minutes % 60
    
    if hours > 0:
        return f"{hours}h {mins}m" if mins > 0 else f"{hours}h"
    else:
        return f"{mins}m"

def minutes_to_decimal_hours(minutes):
    """
    Convert minutes to decimal hours (for calculations)
    Args:
        minutes: Duration in minutes (int or float)
    Returns:
        Decimal hours
    """
    if not minutes or minutes < 0:
        return 0.0
    
    return round(minutes / 60, 2)

File: checkpoints/checkpoint_20251006_140102_start_working_on_the_logbook_review_process_where_the_lgbooks_flows_back_and_forward_between_the_intern_and_supervisor_until_approved/code/frontend/src/utils/durationUtils.ts
/**
 * Utility functions for duration formatting and conversion
 */

/**
 * Convert minutes to hh:mm format
 * @param minutes - Duration in minutes
 * @returns Formatted string like "3:45" or "0:30"
 */
export function minutesToHoursMinutes(minutes: number | string | null | undefined): string {
  const numMinutes = typeof minutes === 'string' ? parseInt(minutes, 10) : Number(minutes)
  
  if (!Number.isFinite(numMinutes) || numMinutes < 0) {
    return '0:00'
  }
  
  // Round to avoid floating point precision issues
  const roundedMinutes = Math.round(numMinutes)
  const hours = Math.floor(roundedMinutes / 60)
  const mins = roundedMinutes % 60
  
  return `${hours}:${mins.toString().padStart(2, '0')}`
}

/**
 * Convert minutes to decimal hours (for calculations)
 * @param minutes - Duration in minutes
 * @returns Decimal hours
 */
export function minutesToDecimalHours(minutes: number | string | null | undefined): number {
  const numMinutes = typeof minutes === 'string' ? parseInt(minutes, 10) : Number(minutes)
  
  if (!Number.isFinite(numMinutes) || numMinutes < 0) {
    return 0
  }
  
  return Math.round((numMinutes / 60) * 10) / 10
}

/**
 * Convert decimal hours to hh:mm format
 * @param decimalHours - Duration in decimal hours (e.g., 3.75)
 * @returns Formatted string like "3:45"
 */
export function decimalHoursToHoursMinutes(decimalHours: number): string {
  if (!Number.isFinite(decimalHours) || decimalHours < 0) {
    return '0:00'
  }
  
  // Round to avoid floating point precision issues
  const totalMinutes = Math.round(decimalHours * 60)
  const hours = Math.floor(totalMinutes / 60)
  const mins = totalMinutes % 60
  
  return `${hours}:${mins.toString().padStart(2, '0')}`
}

/**
 * Format duration with unit (e.g., "3:45h" or "0:30h")
 * @param minutes - Duration in minutes
 * @returns Formatted string with unit
 */
export function formatDurationWithUnit(minutes: number | string | null | undefined): string {
  return `${minutesToHoursMinutes(minutes)}h`
}

/**
 * Format duration for display in cards/summaries (e.g., "3h 45m" or "30m")
 * @param minutes - Duration in minutes
 * @returns Formatted string for display
 */
export function formatDurationDisplay(minutes: number | string | null | undefined): string {
  const numMinutes = typeof minutes === 'string' ? parseInt(minutes, 10) : Number(minutes)
  
  if (!Number.isFinite(numMinutes) || numMinutes < 0) {
    return '0m'
  }
  
  const hours = Math.floor(numMinutes / 60)
  const mins = numMinutes % 60
  
  if (hours > 0) {
    return mins > 0 ? `${hours}h ${mins}m` : `${hours}h`
  } else {
    return `${mins}m`
  }
}
